src/server/network.ts

```diff
@@ -275,15 +294,33 @@ export class Response extends SdkObject {
   private _headersMap = new Map<string, string>();
   private _getResponseBodyCallback: GetResponseBodyCallback;
   private _timing: ResourceTiming;
+  private _connectionDetails: ConnectionDetails;
+  private _connectionDetailsFinishedCallback: (details: ConnectionDetails) => void = _ => {};
+  private _connectionDetailsPromise: Promise<ConnectionDetails>;
 
-  constructor(request: Request, status: number, statusText: string, headers: types.HeadersArray, timing: ResourceTiming, getResponseBodyCallback: GetResponseBodyCallback) {
+  constructor(request: Request, status: number, statusText: string, headers: types.HeadersArray, timing: ResourceTiming, getResponseBodyCallback: GetResponseBodyCallback, connectionDetails: ConnectionDetails & {complete?: boolean}) {
     super(request.frame(), 'response');
     this._request = request;
     this._timing = timing;
     this._status = status;
     this._statusText = statusText;
     this._url = request.url();
     this._headers = headers;
+
+    const filteredConnectionDetails: ConnectionDetails = {};
+    const serverIPAddressAndPort = helper.filterEmpties(connectionDetails.serverIPAddressAndPort);
+    if (serverIPAddressAndPort)
+      filteredConnectionDetails.serverIPAddressAndPort = serverIPAddressAndPort;
+    const securityDetails = helper.filterEmpties(connectionDetails.securityDetails);
+    if (securityDetails)
+      filteredConnectionDetails.securityDetails = securityDetails;
+    this._connectionDetails = filteredConnectionDetails;
+    this._connectionDetailsPromise = new Promise(f => {
+      this._connectionDetailsFinishedCallback = f;
+    });
+    if (connectionDetails.complete ?? true)
```

Comment:
Too fancy! This is a sign that something has gone wrong with the design.

Reply:
Hehe! Totally agree, but might need a few pointers here. For some background:

1. Started simple: just passed in Security Details and Connection Details in create response…but WebKit didn't have this information until _loadingFinished
2. …so tried to move everything there…[but Chromium sometimes doesn't always fire _loadingFinished](https://github.com/microsoft/playwright/pull/6653#discussion_r636250024)…
3. …so arrived at this very awkward design that needs some help 😄 ❤️

Reply:
When chromium doesn't fire loadingFinished, we never dispatch the response event. So I don't think this patch needs to worry about that.

Reply:
Sweet—yeah, I lead myself astray (and to unideal code) by focusing on that test case! Will update.